mathics/core/rules.py

```diff
@@ -62,6 +62,8 @@ def yield_match(vars, rest):
             else:
                 result = new_expression
 
+            if isinstance(result, Expression) and result.elements_properties is None:
```

Comment:
In this method, we apply a rule to an expression to produce a new expression. I guess that in certain cases, the new expression does not need to be reevaluated, and hence, the element_properties that we compute here are not used. Is there a reason to make this here?

Reply:
I first was noticed this empirically - if you don't have it the code will currently crash. And there are discussions in this PR which show why.  We have the setter and there is code that uses that. One of the places where the setter method is used is right here in this method. That could be replaced, but I am leaving that for some other time or some other person. 

The setter-property method sets `self.element_properties` to `None`.  See the discussion above in this PR.

There is a good deal of this code here, that can be improved but it will require other parts of the code to be more sensitive to elements properties. 

That is why we are using a _mutable_ records rather than a named tuple which isn't mutable. 

If it wasn't clear, I repeat that what I have seen in benchmarking is that there are some big bit wins here, especially if you look at an overall sequence of operations such as you'd get by running the entirety of the combinatorica tests. 

But there are also some regressions and a number of those regressions  when you look at specific individual operations. And these can eliminated and some cases turned into even bigger wins after the rest of the code is adjusted to understand that there are elements properties and these other new features using alternative representations (in the case of List) and objects with specialized evaluation functions. 

Right now though we need to introduce some additional and non-optimal checks or some pessimistic behavior to bridge the gaps because we also want _working_ code. 

I  do however have a branch which removes the pessimism and doesn't bridge gaps. I use this for benchmarking performance. Needless to say it is broken in general. When there is something I want to benchmark I incrementally fix the brokenness around that so that I can benchmark some specific feature. And those changes to fix the brokenness then go into master. That is why earlier in the weekend there were what seemed like (and to some extent _were_) random adjustments.

But until  all the gaps are fixed, we have to live with live with some regressions in specific cases, especially when there is a plan in mind for how those regressions will be reversed. 


>  hence, the element_properties that we compute 

Note that I write what in English is awkward "element_properties" rather than the more natural "element_properties" because it is a property of the _all_ elements of the Expression object. And yes, technically an Expression is an Element too, but I want to emphasize the fact these are properties of a _collection_. If folks want to casually refer to this as Element properties where Element refers to the Expression object rather than the elements attribute of the Expression that's okay. For myself, I want to keep emphasizing that this is about elements or the elements field of the Expression element, not a single element in elements.